compiler/src/main/java/com/squareup/anvil/compiler/codegen/AnnotationReferenceExtensions.kt

```diff
@@ -0,0 +1,37 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.squareup.anvil.compiler.internal.argumentType
+import com.squareup.anvil.compiler.internal.classDescriptor
+import com.squareup.anvil.compiler.internal.findAnnotationArgument
+import com.squareup.anvil.compiler.internal.getAnnotationValue
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference.Descriptor
+import com.squareup.anvil.compiler.internal.reference.AnnotationReference.Psi
+import com.squareup.anvil.compiler.internal.reference.AnvilCompilationExceptionAnnotationReference
+import com.squareup.anvil.compiler.internal.reference.ClassReference
+import com.squareup.anvil.compiler.internal.reference.toClassReference
+import com.squareup.anvil.compiler.internal.requireFqName
+import org.jetbrains.kotlin.psi.KtClassLiteralExpression
+import org.jetbrains.kotlin.resolve.constants.KClassValue
+
+internal fun AnnotationReference.parentScope(): ClassReference {
+  return when (this) {
+    is Psi ->
+      annotation
+        .findAnnotationArgument<KtClassLiteralExpression>(
+          name = "parentScope",
+          index = 1
+        )
+        ?.requireFqName(module)
+        ?.toClassReference(module)
+
+    is Descriptor ->
+      (annotation.getAnnotationValue("parentScope") as? KClassValue)
+        ?.argumentType(module)
+        ?.classDescriptor()
+        ?.toClassReference(module)
+  } ?: throw AnvilCompilationExceptionAnnotationReference(
```

Comment:
Is there a use-case where the annotation may not have a parent scope and we'd want to return null instead? Or is the assumption that this will only be called on annotations where one should exist?

Reply:
Yes, the assumption right now is that it's only called for the correct annotations. I thought about introducing subclasses for `AnnotationReference`, but was afraid that it could be much more complicated (another use case is `boundType`). For now I'd leave it as is and if we need an `..orNull()` function, then we can easily add it.